improve `--reduce-test`

alexlamsl opened the pull request:
@kzc this is based on https://github.com/mishoo/UglifyJS2/issues/3738#issuecomment-593701807

Test case from https://github.com/mishoo/UglifyJS2/pull/3716#issuecomment-585554934 shows notable improvement in performance:
#### Before
```js
$ uglifyjs 442602716.js -c --reduce-test
// Node.js v10.19.0 on win32 x64
// reduce test pass 1, iteration 0: 29764 bytes
// reduce test pass 1, iteration 25: 13929 bytes
// reduce test pass 1, iteration 50: 13923 bytes
// reduce test pass 1, iteration 75: 11909 bytes
// reduce test pass 1, iteration 100: 11884 bytes
// reduce test pass 1, iteration 125: 11877 bytes
// reduce test pass 1, iteration 150: 11875 bytes
// reduce test pass 1, iteration 175: 11385 bytes
// reduce test pass 1, iteration 200: 10607 bytes
// reduce test pass 1, iteration 225: 10529 bytes
// reduce test pass 1, iteration 250: 10313 bytes
// reduce test pass 1, iteration 275: 9638 bytes
// reduce test pass 1, iteration 300: 9393 bytes
// reduce test pass 1, iteration 325: 8838 bytes
// reduce test pass 1, iteration 350: 8568 bytes
// reduce test pass 1, iteration 375: 8520 bytes
// reduce test pass 1, iteration 400: 8294 bytes
// reduce test pass 1, iteration 425: 6775 bytes
// reduce test pass 1, iteration 450: 5752 bytes
// reduce test pass 1, iteration 475: 1095 bytes
// reduce test pass 1, iteration 500: 581 bytes
// reduce test pass 1: 567 bytes
// reduce test pass 2: 175 bytes
// reduce test pass 3: 156 bytes
var b = 0;

var expr1 = this;

for (var key1 in expr1) {
    var Infinity_2 = expr1[key1];
    if (Infinity_2) {
        0;
    } else {
        switch (0) {
          case 0:
            --b;
        }
    }
}

while (0) {
    var brake106 = 0;
}

console.log(b);
// output: -2
//
// minify: -3
//
// options: {
//   "compress": true,
//   "mangle": false
// }
```

#### After
```js
$ uglifyjs 442602716.js -c --reduce-test
// Node.js v10.19.0 on win32 x64
// reduce test pass 1, iteration 0: 29764 bytes
// reduce test pass 1, iteration 25: 13929 bytes
// reduce test pass 1, iteration 50: 13923 bytes
// reduce test pass 1, iteration 75: 11909 bytes
// reduce test pass 1, iteration 100: 11884 bytes
// reduce test pass 1, iteration 125: 11413 bytes
// reduce test pass 1, iteration 150: 10858 bytes
// reduce test pass 1, iteration 175: 10824 bytes
// reduce test pass 1, iteration 200: 10305 bytes
// reduce test pass 1, iteration 225: 6565 bytes
// reduce test pass 1, iteration 250: 5471 bytes
// reduce test pass 1, iteration 275: 538 bytes
// reduce test pass 1: 358 bytes
// reduce test pass 2: 137 bytes
// reduce test pass 3: 125 bytes
var expr1 = this;

for (var key1 in expr1) {
    var Infinity_2 = expr1[key1];
    if (Infinity_2) {
        0;
    } else {
        console.log();
    }
}

while (0) {
    var brake106 = 0;
}
// output:
//
// minify:
//
//
// options: {
//   "compress": true,
//   "mangle": false
// }
```

kzc commented:
Cool. Consider emitting the outputs with `JSON.stringify` rather than counting newlines and guessing at spaces:

```patch
--- a/test/reduce.js
+++ b/test/reduce.js
@@ -457,6 +457,6 @@ module.exports = function reduce_test(testcase, minify_options, reduce_options)
         testcase.code += [
             "",
-            "// output: " + to_comment(differs.unminified_result),
-            "// minify: " + to_comment(differs.minified_result),
+            "// output: " + JSON.stringify(differs.unminified_result + ""),
+            "// minify: " + JSON.stringify(differs.minified_result + ""),
             "// options: " + to_comment(minify_options_json),
         ].join("\n").replace(/\u001b\[\d+m/g, "");
```

kzc commented:
Unfortunately the AST_Sub logic omission of this PR make it produce twice as large test cases on average in the xor bug test suite.

Here's a typical result with the latest patch:
```
console.log("function" ^ 0);
// output: "0\n"
// minify: "function0\n"
```
versus this PR (with outputs stringified):
```
[ a_1 = ("function" ^ 1) & 1, a_1 && (c = 0) ];

c;
// output: ""
// minify: "ReferenceError: c is not defined"
```
By the way, if you add the following to the top of `test/reduce.js` and ufuzz you won't have to deal with the tty escape codes:
```
process.env.NODE_DISABLE_COLORS=1
```
Node v10 does not respect NO_COLOR=1 unfortunately.  NODE_DISABLE_COLORS works on all Node versions starting from v10.

kzc commented:
> Unfortunately the AST_Sub logic omission of this PR make it produce twice as large test cases on average in the xor bug test suite.

I was mistaken - the PR was not correctly applied in the testing in my previous post.

This PR produces near equivalent reduced results as my patch for the xor bug test suite.

kzc commented on test/reduce.js:
```diff
@@ -547,6 +570,7 @@ function producesDifferentResultWhenMinified(result_cache, code, minify_options,
     var elapsed = Date.now();
     var unminified_result = run_code(result_cache, code, toplevel, max_timeout);
     elapsed = Date.now() - elapsed;
+    if (typeof unminified_result == "string" && unminified_result.split(/\n/, 10) == 10) return false;
```
Certainly better than my regex. But was `.length` omitted by mistake?
```
$ node -p '"ABCBACBABCBAAB".split(/A/, 4)'
[ '', 'BCB', 'CB', 'BCB' ]
$ node -p '"ABCBACBABCBAAB".split(/A/, 4) == 4'
false
$ node -p '"ABCBACBABCBAAB".split(/A/, 4).length == 4'
true
```

kzc replied:
Unrelated to the possible typo above... While this logic that I had suggested limiting result lines to 10 would be fine for ufuzz cases, it might be too restrictive for regular user reported `console.log` test cases and might result in "```Can't reproduce test failure with minify options provided```". Could a check be made on the number of lines that the original (first) unminified version of the code emits and use *that* limit here rather than a hardcoded number?

alexlamsl replied:
Oops! 😅

alexlamsl replied:
Fixed in https://github.com/mishoo/UglifyJS2/pull/3742/commits/30a88298e76c42794812a551bfbb92b7ad396d68

alexlamsl replied:
(Sorry, Internet dropped out... again.)

I replied before seeing your latest message &minus; I'm currently working your suggestion, though I hit onto something that I need to debug.

alexlamsl commented:
> Cool. Consider emitting the outputs with JSON.stringify rather than counting newlines and guessing at spaces:

After https://github.com/mishoo/UglifyJS2/pull/3742/commits/bda6a72ddc6b9663cf5eaa28a035f8cf06ed8320:
```js
// Node.js v10.19.0 on win32 x64
// reduce test pass 1, iteration 0: 29764 bytes
// reduce test pass 1, iteration 25: 13929 bytes
// reduce test pass 1, iteration 50: 13923 bytes
// reduce test pass 1, iteration 75: 11909 bytes
// reduce test pass 1, iteration 100: 11884 bytes
// reduce test pass 1, iteration 125: 11413 bytes
// reduce test pass 1, iteration 150: 10858 bytes
// reduce test pass 1, iteration 175: 10824 bytes
// reduce test pass 1, iteration 200: 10305 bytes
// reduce test pass 1, iteration 225: 6565 bytes
// reduce test pass 1, iteration 250: 5471 bytes
// reduce test pass 1, iteration 275: 538 bytes
// reduce test pass 1: 358 bytes
// reduce test pass 2: 137 bytes
// reduce test pass 3: 125 bytes
var expr1 = this;

for (var key1 in expr1) {
    var Infinity_2 = expr1[key1];
    if (Infinity_2) {
        0;
    } else {
        console.log();
    }
}

while (0) {
    var brake106 = 0;
}
// (stringified)
// output: "\n"
// minify: "\n\n"
// options: {
//   "compress": true,
//   "mangle": false
// }
```

kzc commented:
> After bda6a72:

That does the trick, but since I discovered `process.env.NODE_DISABLE_COLORS = 1;`, I use it in all my console apps. The new NodeJS `console.log` tty coloring is quite annoying and should have been an opt-in rather than an opt-out.

alexlamsl commented:
> The new NodeJS `console.log` tty coloring is quite annoying and should have been an opt-in rather than an opt-out.

I concur &minus; though I've been dealing with many other aspects of this platform (hint: `npm`) that this particular issue is of relatively minor concern.

I do know about the configuration flag, it's just that in general I don't like to rely on global environment variables if I can avoid it. The current regex patches it up nicely for all foreseeable bugs/features from Node.js, so I prefer it then trusting the underlying runtime to behave 😉

kzc commented:
Any thoughts on https://github.com/mishoo/UglifyJS2/pull/3742#discussion_r387086450 ?

kzc commented:
> in general I don't like to rely on global environment variables if I can avoid it. The current regex patches it up nicely for all foreseeable bugs/features from Node.js, so I prefer it then trusting the underlying runtime to behave

While I don't like to set environment variables (even for just the process lifetime), it's less code, less complexity, and unfortunately the safest way to guard against any future escape code tty nonsense.

kzc commented:
In my testing of this AST_Assign permutation I only saw an 8% speed improvement reducing xor bugs, but I see that the test case in the top post has half the number of iterations with this approach. I wonder if this is typical of most reduced test cases or just an outlier.

alexlamsl commented:
> I see that the test case in the top post has half the number of iterations with this approach. I wonder if this is typical of most reduced test cases or just an outlier.

I'd say it's typical, from my experience of manually reducing those test cases in the past.

alexlamsl commented:
Wow so after I've made that line counting work properly it's actually slower now, without giving any better results:
```js
// Node.js v10.19.0 on win32 x64
// reduce test pass 1, iteration 0: 29764 bytes
// reduce test pass 1, iteration 25: 13929 bytes
// reduce test pass 1, iteration 50: 13923 bytes
// reduce test pass 1, iteration 75: 11909 bytes
// reduce test pass 1, iteration 100: 11884 bytes
// reduce test pass 1, iteration 125: 11877 bytes
// reduce test pass 1, iteration 150: 11875 bytes
// reduce test pass 1, iteration 175: 11385 bytes
// reduce test pass 1, iteration 200: 10607 bytes
// reduce test pass 1, iteration 225: 10529 bytes
// reduce test pass 1, iteration 250: 10313 bytes
// reduce test pass 1, iteration 275: 9638 bytes
// reduce test pass 1, iteration 300: 9393 bytes
// reduce test pass 1, iteration 325: 8838 bytes
// reduce test pass 1, iteration 350: 8568 bytes
// reduce test pass 1, iteration 375: 8520 bytes
// reduce test pass 1, iteration 400: 8294 bytes
// reduce test pass 1, iteration 425: 6775 bytes
// reduce test pass 1, iteration 450: 5752 bytes
// reduce test pass 1, iteration 475: 1095 bytes
// reduce test pass 1, iteration 500: 581 bytes
// reduce test pass 1: 567 bytes
// reduce test pass 2: 175 bytes
// reduce test pass 3: 156 bytes
var b = 0;

var expr1 = this;

for (var key1 in expr1) {
    var Infinity_2 = expr1[key1];
    if (Infinity_2) {
        0;
    } else {
        switch (0) {
          case 0:
            --b;
        }
    }
}

while (0) {
    var brake106 = 0;
}

console.log(b);
// output: -2
//
// minify: -3
//
// options: {
//   "compress": true,
//   "mangle": false
// }
```

kzc commented:
> Wow so after I've made that line counting work properly it's actually slower now, without giving any better results:

That's strange. What's expensive about counting the number of newlines for the first unminified run?

alexlamsl commented:
Tweaking the overflow logic to report instead of skipping the test case:
```diff
--- a/test/reduce.js
+++ b/test/reduce.js
@@ -589,7 +589,7 @@ function producesDifferentResultWhenMinified(result_cache, code, minify_options,
     var elapsed = Date.now();
     var unminified_result = run_code(result_cache, code, toplevel, max_timeout);
     elapsed = Date.now() - elapsed;
-    if (typeof unminified_result == "string" && unminified_result.split(/\n/, 10).length == 10) return false;
+    if (typeof unminified_result == "string" && unminified_result.split(/\n/, 100).length == 100) console.error(unminified_result.split(/\n/).length);
     var timeout = Math.min(100 * elapsed, max_timeout);
     var minified_result = run_code(result_cache, minified.code, toplevel, timeout);
```
```js
// Node.js v10.19.0 on win32 x64
// reduce test pass 1, iteration 0: 29764 bytes
// reduce test pass 1, iteration 25: 13929 bytes
// reduce test pass 1, iteration 50: 13923 bytes
// reduce test pass 1, iteration 75: 11909 bytes
// reduce test pass 1, iteration 100: 11884 bytes
162
162
162
157
157
157
157
157
// reduce test pass 1, iteration 125: 11413 bytes
167
167
167
167
167
167
167
167
167
152
147
147
147
147
147
147
147
147
// reduce test pass 1, iteration 150: 10858 bytes
147
152
152
152
152
152
152
152
// reduce test pass 1, iteration 175: 10824 bytes
152
152
147
147
152
152
147
147
152
152
152
// reduce test pass 1, iteration 200: 10305 bytes
152
152
152
152
152
147
122
122
117
117
117
117
117
117
117
117
112
112
// reduce test pass 1, iteration 225: 6565 bytes
112
112
112
112
112
// reduce test pass 1, iteration 250: 5471 bytes
// reduce test pass 1, iteration 275: 538 bytes
// reduce test pass 1: 358 bytes
// reduce test pass 2: 137 bytes
// reduce test pass 3: 125 bytes
var expr1 = this;

for (var key1 in expr1) {
    var Infinity_2 = expr1[key1];
    if (Infinity_2) {
        0;
    } else {
        console.log();
    }
}

while (0) {
    var brake106 = 0;
}
// (stringified)
// output: "\n"
// minify: "\n\n"
// options: {
//   "compress": true,
//   "mangle": false
// }
```

alexlamsl commented:
Looks like even 100 lines was too conservative, let along 10 👻

kzc commented:
The line counting thing is just a nice to have for the CLI. For ufuzz it's fine with the hardcoded 10. 

Unrelated... I just realized that the `console.log` node creation logic would apply equally well to AST_Binary as it does for AST_Assign, as the latter is derived from the former.

If you get rid the AST_Assign logic altogether and use the following it'll produce the same results:

```
            else if (node instanceof U.AST_Binary) {
                var mod = ((node.start._permute += step) * steps | 0) % 4;
                var expr = [
                    node.left,
                    node.right,
                    null,  // wrap node.right with console.log
                    null,  // wrap node.left with console.log
                ][ mod ];
                if (!expr) {
                    expr = new U.AST_Call({
                        expression: new U.AST_Dot({
                            expression: new U.AST_SymbolRef({
                                name: "console",
                                start: {},
                            }),
                            property: "log",
                            start: {},
                        }),
                        args: [ (mod & 1) ? node.left : node.right ],
                        start: {},
                    });
                }
                CHANGED = true;
                return expr;
            }
```
The timings are comparable, and it produces slightly better results for xor bug test suite.

alexlamsl commented:
Here's an interesting test case between Node.js 10 and 12, in both `sandbox.run_code()` and `cat | node`:
```js
var d, a = this;
for (var b in a) {
    if (b.length > 1) continue;
    console.log(b, "" + c);
    c && console.log();
    var c = a[b];
}
```
<table>
<tr>
<th>Node.js</th>
<th>

`run_code()`</th>
<th>

`cat | node`</th>
</tr>
<tr>
<td>10.19.0</td>
<td>

```
a undefined
d [object Object]

c undefined
b undefined
```
</td>
<td>

```
d undefined
a undefined
c [object global]

b [object global]
```
</td>
</tr>
<tr>
<td>12.16.1</td>
<td>

```
a undefined
d [object Object]

b undefined
c b
```
</td>
<td>

```
d undefined
a undefined
b [object global]

c b
```
</td>
</tr>
</table>

kzc commented:
If the chart had two columns for run_code for toplevel true and false respectively, what are the results?

alexlamsl commented:
Both `.run_code(test, true)` and `node test.js` returns `""` on both versions of Node.js

kzc commented:
What happened to the AST_Assign/console.log logic?

kzc commented:
No interest in latest variation https://github.com/mishoo/UglifyJS2/pull/3742#issuecomment-594116907 ?

alexlamsl commented:
Those changes were causing real test failure in `mocha` (the "reasonable performance" one) so I've took them out for now &minus; still in the middle of getting the new correctness tests to work, will revisit those code-insertion changes once that's sorted out.

kzc commented:
I think the only point of contention was the number of lines logic. It may be just a degenerate case. We can nix the counting the lines stuff and hardcode 10 for ufuzz, which ran just fine.

I get considerably better performance with the new AST_Binary/console.log patch above and better reduced results for most tests. It's up to you. I'll keep using it in my fork.

alexlamsl commented:
Right I've incorporated a tweaked version of https://github.com/mishoo/UglifyJS2/pull/3742#issuecomment-594116907 into the latest push and updated the infinite loop test accordingly &minus; will merge if the tests pass.

kzc commented:
Whatever works for you. Reducing tests was an interesting puzzle. I don't see any need for further refinements.

alexlamsl commented:
It certainly helps in filtering out false positives, now that we have fewer bugs.

kzc commented:
The `AST_Binary` `console.log` addition is working well:
```
2020-03-21T19:01:43.8633130Z // reduced test case (output will differ)
2020-03-21T19:01:43.8633360Z 
2020-03-21T19:01:43.8633737Z // (beautified)
2020-03-21T19:01:43.8634564Z console.log(1 - .1 - 1);
2020-03-21T19:01:43.8635001Z // output: -0.09999999999999998
2020-03-21T19:01:43.8635262Z // 
2020-03-21T19:01:43.8635967Z // minify: -0.1
2020-03-21T19:01:43.8636220Z // 
2020-03-21T19:01:43.8636454Z // options: {
2020-03-21T19:01:43.8636692Z //   "compress": {
2020-03-21T19:01:43.8636925Z //     "keep_fargs": false,
2020-03-21T19:01:43.8637164Z //     "passes": 1000000,
2020-03-21T19:01:43.8637407Z //     "sequences": 1000000,
2020-03-21T19:01:43.8637647Z //     "unsafe": true,
2020-03-21T19:01:43.8637868Z //     "unsafe_Function": true,
2020-03-21T19:01:43.8638107Z //     "unsafe_math": true,
2020-03-21T19:01:43.8638341Z //     "unsafe_proto": true,
2020-03-21T19:01:43.8638584Z //     "unsafe_regexp": true
2020-03-21T19:01:43.8638820Z //   }
2020-03-21T19:01:43.8639219Z // }
```
